internal/expr/scan.go

```diff
@@ -148,28 +160,54 @@ func (re *ResultExpr) decodeValue(v reflect.Value) error {
 	return nil
 }
 
-func setValue(dest reflect.Value, fInfo field, val any) error {
+func setValue(dest reflect.Value, fInfo fielder, val any) error {
 	var isZero bool
 
 	v := reflect.ValueOf(val)
+	name := fInfo.Name()
+
+	fmt.Printf("%s, %v\n", name, dest.Type().Kind())
 
-	if val == nil {
-		if fInfo.omitEmpty {
-			return nil
+	switch dest.Type().Kind() {
+	case reflect.Struct:
+		f, ok := fInfo.(field)
+		if !ok {
+			return fmt.Errorf("internal error: argument of type %#v has no member of type %#v", dest.Type(), f)
 		}
-		isZero = true
-		v = reflect.Zero(fInfo.typ)
-	}
 
-	if !isZero && v.Type() != fInfo.typ {
-		return fmt.Errorf("result of type %#v but field %#v is type %#v", v.Type().Name(), fInfo.name, fInfo.typ.Name())
-	}
-	f := dest.FieldByIndex(fInfo.index)
-	if !f.CanSet() {
-		return fmt.Errorf("cannot set field %#v. CanAddr=%v", fInfo.name, f.CanAddr())
+		if val == nil {
+			if f.omitEmpty {
+				return nil
+			}
+			isZero = true
+			v = reflect.Zero(f.typ)
+		}
+		if !isZero && v.Type() != f.typ {
+			return fmt.Errorf("result of type %#v but field %#v is type %#v", v.Type().Name(), name, f.typ.Name())
+		}
+		itsField := dest.FieldByIndex(f.index)
+		if !itsField.CanSet() {
+			return fmt.Errorf("cannot set field %#v. CanAddr=%v", f.name, itsField.CanAddr())
+		}
+		itsField.Set(v)
+		return nil
+	case reflect.Map:
+		m, ok := fInfo.(mapKey)
```

Comment:
Is the m of type `mapKey` actually needed explicitly here, we already have the `name` which is its only field.

Reply:
Yeah could omit `m` here. Was meaning to keep the two cases of the switch statement similar, and also referring to `m` in the error message below, but maybe we could replace that with something else. This method still needs some work, as one of the test functions currently fail. Will update once that issue has been resolved.